File: packages/core/opensession-server/src/frontend/lib/slides-block.ts
/**
 * The cheap half of a slides block: which fences are decks, and how a deck's
 * markdown splits into slides. Building the deck (slides-deck.ts) pulls in
 * the markdown renderer, so it is imported when a body carries one.
 */

import type { FenceUpgrader } from "./fence-upgraders";

const FENCE_OPEN = /^ {0,3}(`{3,}|~{3,})/;

/**
 * Split on lines that are exactly `---` (trailing whitespace allowed). A
 * `---` inside a nested code fence belongs to that fence, not the deck.
 * Blank slides are dropped, so a deck that opens with `---` does not start
 * on an empty one.
 */
export function splitSlides(source: string): string[] {
  const slides: string[] = [];
  let current: string[] = [];
  let fence: string | null = null;
  for (const line of source.split(/\r?\n/)) {
    const open = FENCE_OPEN.exec(line);
    if (fence) {
      const closes =
        open &&
        open[1][0] === fence[0] &&
        open[1].length >= fence.length &&
        line.trim() === open[1];
      if (closes) fence = null;
      current.push(line);
      continue;
    }
    if (open) {
      fence = open[1];
      current.push(line);
      continue;
    }
    if (line.trimEnd() === "---") {
      slides.push(current.join("\n"));
      current = [];
      continue;
    }
    current.push(line);
  }
  slides.push(current.join("\n"));
  return slides.map((slide) => slide.trim()).filter(Boolean);
}

let deckPromise: Promise<typeof import("./slides-deck")> | null = null;
function loadDeck() {
  deckPromise ??= import("./slides-deck");
  return deckPromise;
}

/** A fence with nothing in it keeps the plain code block. */
export const slidesUpgrader: FenceUpgrader = {
  langs: ["slides"],
  async upgrade({ pre, source, root, alive, markdown }) {
    const slides = splitSlides(source);
    if (slides.length === 0) return false;
    const m = await loadDeck().catch(() => null);
    if (!m || !alive() || !root.contains(pre)) return false;
    pre.replaceWith(
      m.buildSlidesDeck(slides, { expandable: true, markdown }).el,
    );
    return true;
  },
};

File: packages/clients/ios/OS1/Models/SlidesBlock.swift
import Foundation

/// A ```slides fence is markdown split into slides on lines that are exactly
/// `---` (docs/blocks.md, "Slides").
enum SlidesBlock {
    /// Split on lines that are exactly `---` (trailing whitespace allowed). A
    /// `---` inside a nested code fence belongs to that fence, not the deck.
    /// Blank slides are dropped, so a deck that opens with `---` does not
    /// start on an empty one.
    static func split(_ source: String) -> [String] {
        var slides: [String] = []
        var current: [Substring] = []
        var fence: (marker: Character, length: Int)?
        for line in source.split(separator: "\n", omittingEmptySubsequences: false) {
            let line = line.last == "\r" ? line.dropLast() : line
            if let open = fence {
                if MarkdownFenceSegmenter.closes(line, marker: open.marker, length: open.length) {
                    fence = nil
                }
                current.append(line)
                continue
            }
            if let opened = MarkdownFenceSegmenter.opening(line) {
                fence = (opened.marker, opened.length)
                current.append(line)
                continue
            }
            var trimmed = line
            while let last = trimmed.last, last == " " || last == "\t" { trimmed = trimmed.dropLast() }
            if trimmed == "---" {
                slides.append(current.joined(separator: "\n"))
                current = []
                continue
            }
            current.append(line)
        }
        slides.append(current.joined(separator: "\n"))
        return slides
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
    }
}
